cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+                        // such thing at this late stage in the pipeline.
+                        panic!("cleanup_ebb: spill with no offset");
+                    }
+                }
+                InstructionData::Unary {
+                    opcode: Opcode::Fill,
+                    arg,
+                } => {
+                    // Extract: (dst_value, dst_regunit, offset, size)
+                    // Invalidate: (offset, size)
+                    // Add new: {dst_regunit -> (dst_value, offset, size)}
+                    let results = self.cur.func.dfg.inst_results(inst);
+                    let slot = self.slot_of_value(*arg);
+                    let src_value = arg;
+                    let dst_value = results[0];
+                    let dst_reg = self.reg_of_value(dst_value);
+                    if let Some(offs) = slot.offset {
+                        let size = slot.size;
+                        if env.has_exact_binding(dst_reg, offs, size) {
+                            // This instruction is an exact copy of a fill we
+                            // saw earlier, and the loaded value is still
+                            // valid.  So we'll schedule this instruction for
+                            // deletion (below).
+                            replacement = Some((true, *src_value));
+                        } else if let Some((reg2, val2)) =
+                            env.has_inexact_binding(self.reginfo.banks, dst_reg, offs, size)
+                        {
+                            debug_assert!(reg2 != dst_reg);
+                            replacement = Some((false, val2));
+                            // This fill is from the required slot, but into a
+                            // different register `reg2`.  So replace it with
+                            // a copy from `reg2` to `dst_reg` and update
+                            // `dst_reg`s binding to make it the same as
+                            // `reg2`'s, so as to maximise the chances of
+                            // future matches after this instruction.
+                            env.bind(dst_reg, dst_value, offs, size);
+                        } else {
+                            // This fill creates some new binding we don't
+                            // know about.  Update `env` to track it.
+                            env.invalidate_by_offset(offs, size);
+                            env.bind(dst_reg, dst_value, offs, size);
+                        }
+                    } else {
+                        // A fill insn with no offset.
+                        panic!("cleanup_ebb: fill with no offset");
+                    }
+                }
+                InstructionData::RegSpill { .. }
+                | InstructionData::RegFill { .. }
+                | InstructionData::Call { .. }
+                | InstructionData::CallIndirect { .. } => {
+                    // Play safe: forget everything we know.
```

Comment:
Why would we need to forget everything after a call or call_indirect? Is that because a call might write on the caller's stack (and thus fill a slot in a way we're not tracking)?

Reply:
In general, if calls were correctly annotated with all (or, more generally, a superset of) the registers they write, directly or indirectly, then we wouldn't need to special-case them, since the general-case logic at the end of this instruction-handling `match` would handle them.  (This relates to an email conversion about calls that happened early in your/Lars'/my engagement with Cranelift).  However, AFAIK this isn't the case, so the simple-if-suboptimal thing is to forget everything.  Given that few of the hot blocks that I've studied so far contain calls or returns, this doesn't strike me as much of a loss.

I don't *think* that a call can overwrite our own stack slots (that would seem pretty strange to me), but I'm not 100%.  So forgetting everything is also the safe option from this point of view.